network-mux/src/Network/Mux/Bearer/Socket.hs

```diff
@@ -52,25 +54,34 @@ socketAsMuxBearer sd = do
     where
       readSocket :: (HasCallStack) => IO (Mx.MuxSDU ptcl, Time IO)
       readSocket = do
-          hbuf <- recvLen' 8 []
+          hbuf <- recvLen' True 8 []
           --say "read"
           --hexDump hbuf ""
           case Mx.decodeMuxSDUHeader hbuf of
               Left  e      -> throwM e
               Right header -> do
                   -- say $ printf "decoded mux header, goint to read %d bytes" (Mx.msLength header)
-                  blob <- recvLen' (fromIntegral $ Mx.msLength header) []
+                  blob <- recvLen' False (fromIntegral $ Mx.msLength header) []
                   ts <- getMonotonicTime
                   --hexDump blob ""
                   return (header {Mx.msBlob = blob}, ts)
 
-      recvLen' :: Int64 -> [BL.ByteString] -> IO BL.ByteString
-      recvLen' 0 bufs = return (BL.concat $ reverse bufs)
-      recvLen' l bufs = do
+      recvLen' :: Bool -> Int64 -> [BL.ByteString] -> IO BL.ByteString
+      recvLen' _ 0 bufs = return (BL.concat $ reverse bufs)
+      recvLen' waitingOnNxtHeader l bufs = do
           buf <- Socket.recv sd l
           if BL.null buf
-              then throwM $ Mx.MuxError Mx.MuxBearerClosed (show sd ++ " closed when reading data") callStack
-              else recvLen' (l - fromIntegral (BL.length buf)) (buf : bufs)
+              then do
+                  when (waitingOnNxtHeader) $
+                      {- This may not be an error, but could be an orderly shutdown.
+                       - We wait 5 seconds to give the mux protocols time to perform
+                       - a clean up and exit.
+                       -}
+                      threadDelay 5
+                  throwM $ Mx.MuxError Mx.MuxBearerClosed (show sd ++
+                      " closed when reading data, waiting on next header " ++
+                      show waitingOnNxtHeader) callStack
```

Comment:
@dcoutts this seems for me as a missing feature.  After reading `MsgDone`, we should not listen any more on incoming messages. In more traditional session type frameworks, it is the channel that is changing the state, and it is easier to encode this kind of information: after reading `MsgDone` the channel is not readable any more.

`5` seconds might be too much (in `praos` slots will have `2` seconds, though a node will be able to skip some), what about putting here just `1` second delay. I guess we are risking that a busy node will error, but a proper fix in my opinion should be done in `typed-protocols` and this is just a temporary fix.

Reply:
Giving it a second thought, I don't think we can address this in `typed-protocols`. The main problem here is that the mux is reading in a loop, and these reads are not orchestrated by the protocols that are multiplexed.

Another observation: since `demux` is eagerly reading we are not creating back-pressure.

Reply:
After receiving `MsgDone` we will stop listening for incoming messages, but this is a multithreaded application and it is another thread that will recieive `MsgDone` so it will take some time before it can signal that we should stop listening (by exiting). 

The 5s wait will _only_ happen if it takes more than 5s for the minprotocol threads to cleanup after them self, but lowering it to 1s wouldn't be the end of the world.

As for a fix in `typed-protocols` for this, this is a muxbearer used by multiple typed-protocols, I don't think that type-protocols can cope with that.

Reply:
I agree, we'd  need a multiplexer which is aware of the states of all the miniprotocols.